### PR TITLE
First pass at adding encrypted streams.  This uses openssl as

### PR DESCRIPTION
**Pull request by kithrup**
installed on FreeBSD, this may need to change.  (The path, I mean.)
It also currently hardcodes the encryption scheme, which should also
change.

### CODE REVIEW

File: z3/snap.py [hunk 1]
```diff
@@ -346,11 +364,13 @@ def backup_full(self, snap_name=None, dry_run=False):
                 capture=True))
         self._cmd.pipe(
             "zfs send '{}'".format(z_snap.name),
-            self._compress(
-                self._pput_cmd(
-                    estimated=estimated_size,
-                    s3_prefix=self.s3_manager.s3_prefix,
-                    snap_name=z_snap.name)
+            self._encrypt(
```
Comment (rciorba) [inline z3/snap.py]: :+1: The order is good, compress then encrypt.

**Comment by kithrup** [comment]
I fixed the two things above in my branch (the HOST one was supposed to be removed, it doesn't work with boto2).

**Comment by benyanke** [comment]
Is there any progress on this? This tool seems awesome, but lack of encryption is a dealbreaker at the moment.

File: z3/snap.py (hunk 1 above)
Comment (ElvishJerricco) [reply z3/snap.py]: @rciorba I don't think the order is correct. `self._encrypt` takes as an argument the command which should *follow* encryption. The argument to `self._encrypt` here is the compression command; hence compression is *following* encryption.

**Pull request merged**